Injecting Views in Parent Class throws IllegalAccessError

I've got a super class in `a.b.widget` and a child class in `a.b.something`. If the fields are set to anything but public an IllegalAccessError is thrown.

```
        java.lang.RuntimeException: Unable to start activity ComponentInfo{com.squareup.cardcase.development/com.squareup.cardcase.home.WalletHomeActivity}: butterknife.Views$UnableToInjectException: Unable to inject views for com.squareup.cardcase.home.ActivityHistoryPanel{4176d420 V.E..... ......I. 0,0-0,0}
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2180)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2230)
        at android.app.ActivityThread.access$600(ActivityThread.java:141)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1234)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:137)
        at android.app.ActivityThread.main(ActivityThread.java:5041)
        at java.lang.reflect.Method.invokeNative(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:511)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:793)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:560)
        at dalvik.system.NativeStart.main(Native Method)
        Caused by: butterknife.Views$UnableToInjectException: Unable to inject views for com.squareup.cardcase.home.ActivityHistoryPanel{4176d420 V.E..... ......I. 0,0-0,0}
        at butterknife.Views.inject(Views.java:125)
        at butterknife.Views.inject(Views.java:77)
        at com.squareup.cardcase.ui.widget.MessageAndListView.<init>(MessageAndListView.java:22)
        at com.squareup.cardcase.home.ActivityHistoryPanel.<init>(ActivityHistoryPanel.java:37)
        at com.squareup.cardcase.home.HomeFragment.buildView(HomeFragment.java:46)
        at com.squareup.cardcase.ui.AbstractWalletFragment.onCreateView(AbstractWalletFragment.java:48)
        at android.app.Fragment.performCreateView(Fragment.java:1695)
        at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:885)
        at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1057)
        at android.app.BackStackRecord.run(BackStackRecord.java:682)
        at android.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1435)
        at android.app.Activity.performStart(Activity.java:5113)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2153)
        ... 11 more
        Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Method.invokeNative(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:511)
        at butterknife.Views.inject(Views.java:117)
        ... 23 more
        Caused by: java.lang.IllegalAccessError: tried to access field com.squareup.cardcase.home.ActivityHistoryPanel.messageSheet from class com.squareup.cardcase.home.ActivityHistoryPanel$$ViewInjector
        at com.squareup.cardcase.home.ActivityHistoryPanel$$ViewInjector.inject(ActivityHistoryPanel$$ViewInjector.java from OutputFileObject:8)
        ... 26 more
```

--- a/butterknife/src/it/superclass/pom.xml
+++ b/butterknife/src/it/superclass/pom.xml
@@ -1,0 +1,44 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<project
+    xmlns="http://maven.apache.org/POM/4.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>com.example.butterknife.tests</groupId>
+  <artifactId>superclass</artifactId>
+  <version>HEAD-SNAPSHOT</version>
+
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+  </properties>
+
+  <dependencies>
+    <dependency>
+      <groupId>com.google.android</groupId>
+      <artifactId>android</artifactId>
+      <version>4.1.1.4</version>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>com.jakewharton</groupId>
+      <artifactId>butterknife</artifactId>
+      <version>@project.version@</version>
+    </dependency>
+  </dependencies>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-compiler-plugin</artifactId>
+        <version>3.0</version>
+        <configuration>
+          <source>1.5</source>
+          <target>1.5</target>
+        </configuration>
+      </plugin>
+    </plugins>
+  </build>
+</project>

--- a/butterknife/src/it/superclass/src/main/java/bar/Test.java
+++ b/butterknife/src/it/superclass/src/main/java/bar/Test.java
@@ -1,0 +1,12 @@
+package bar;
+
+import android.app.Activity;
+import android.view.View;
+import butterknife.InjectView;
+import foo.BaseTest;
+
+public class Test extends BaseTest.Thing {
+  @InjectView(4) View thing4;
+  @InjectView(5) View thing5;
+  @InjectView(6) View thing6;
+}

--- a/butterknife/src/it/superclass/src/main/java/foo/BaseTest.java
+++ b/butterknife/src/it/superclass/src/main/java/foo/BaseTest.java
@@ -1,0 +1,13 @@
+package foo;
+
+import android.app.Activity;
+import android.view.View;
+import butterknife.InjectView;
+
+public class BaseTest {
+  public static class Thing extends Activity {
+    @InjectView(1) View thing1;
+    @InjectView(2) View thing2;
+    @InjectView(3) View thing3;
+  }
+}

--- a/butterknife/src/main/java/butterknife/Views.java
+++ b/butterknife/src/main/java/butterknife/Views.java
@@ -220,25 +220,24 @@ public class Views {
         Set<InjectionPoint> injectionPoints = injection.getValue();
 
         String targetType = type.getQualifiedName().toString();
-        String packageName = elementUtils.getPackageOf(type).getQualifiedName().toString();
-        String className =
-            type.getQualifiedName().toString().substring(packageName.length() + 1).replace('.', '$')
-                + SUFFIX;
-        String parentClass = resolveParentType(type, injectionTargets);
-        StringBuilder injections = new StringBuilder();
-        if (parentClass != null) {
-          injections.append(String.format(PARENT, parentClass)).append('\n');
+        String classPackage = getPackageName(type);
+        String className = getClassName(type, classPackage) + SUFFIX;
+        String classFqcn = classPackage + "." + className;
+        String parentClassFqcn = findParentFqcn(type, injectionTargets);
+        StringBuilder injectionBuilder = new StringBuilder();
+        if (parentClassFqcn != null) {
+          injectionBuilder.append(String.format(PARENT, parentClassFqcn, SUFFIX)).append('\n');
         }
         for (InjectionPoint injectionPoint : injectionPoints) {
-          injections.append(injectionPoint).append('\n');
+          injectionBuilder.append(injectionPoint).append('\n');
         }
+        String injections = injectionBuilder.toString();
 
         // Write the view injector class.
         try {
-          JavaFileObject jfo = filer.createSourceFile(packageName + "." + className, type);
+          JavaFileObject jfo = filer.createSourceFile(classFqcn, type);
           Writer writer = jfo.openWriter();
-          writer.write(
-              String.format(INJECTOR, packageName, className, targetType, injections.toString()));
+          writer.write(String.format(INJECTOR, classPackage, className, targetType, injections));
           writer.flush();
           writer.close();
         } catch (IOException e) {
@@ -250,18 +249,30 @@ public class Views {
     }
 
     /** Finds the parent injector type in the supplied set, if any. */
-    private String resolveParentType(TypeElement typeElement, Set<TypeMirror> parents) {
+    private String findParentFqcn(TypeElement typeElement, Set<TypeMirror> parents) {
       TypeMirror type;
       while (true) {
         type = typeElement.getSuperclass();
         if (type.getKind() == TypeKind.NONE) {
           return null;
         }
-        if (parents.contains(type)) {
-          return type.toString();
-        }
         typeElement = (TypeElement) ((DeclaredType) type).asElement();
+        if (parents.contains(type)) {
+          String packageName = getPackageName(typeElement);
+          return packageName + "." + getClassName(typeElement, packageName);
+        }
       }
+    }
+
+    private String getPackageName(TypeElement type) {
+      return processingEnv.getElementUtils().getPackageOf(type).getQualifiedName().toString();
+    }
+
+    private static String getClassName(TypeElement type, String packageName) {
+      return type.getQualifiedName()
+          .toString()
+          .substring(packageName.length() + 1)
+          .replace('.', '$');
     }
 
     private static class InjectionPoint {
@@ -279,7 +290,7 @@ public class Views {
     }
 
     private static final String INJECTION = "    target.%s = finder.findById(source, %s);";
-    private static final String PARENT = "    %s.inject(finder, target, source);";
+    private static final String PARENT = "    %s%s.inject(finder, target, source);";
     private static final String INJECTOR = ""
         + "// Generated code from Butter Knife. Do not modify!\n"
         + "package %s;\n\n"

